src/cdn_definitions/data.yaml

```diff
@@ -251,3 +251,61 @@ repo_overrides:
       if_created_after: "2023-06-01T14:00:00Z"
       key: example
       value: true
+
+# directory_layouts is used to define DirectoryLayout objects and consist of a
+# pattern, to match against content URL, and attributes for the resulting
+# object, often present as regex groups in the pattern.
+#
+# Allowed value types; string, bool, null
+#
+# With the exception of "rhelalt", which is expected to be stored as bool,
+# a "true" value is used to signal that regex matching should be applied for the
+# corresponding group in the pattern.
+#
+# E.g.;
+#   `type: true`    ==  `type = re.match(...).group("type")`
+#   `type: false`   ==  `type = False`
+#   `type: "dist"`  ==  `type = "dist"`
+directory_layouts:
+  - name: DirectoryLayoutC
+    # "/content/<type>/<platform>/<rhui>/<variant>/<major>.<minor>/<arch>/<rest>"
+    pattern: "/(?:content|shadow)/(?P<type>aus|dist|beta|(?:retired-)?els|eus|htb|rhb|tus|e4s)/(?P<platform>rhes|rhs|cf-me)/(?P<rhui>rhui/)?(?P<variant>[^/]+)/(?P<major_version>[0-9])\\.(?P<minor_version>[0-9]+)/(?P<arch>[^/]+)/(?P<rest>.+)$"
+    type: true
+    platform: true
```

Comment:
I'm not sure that I'm understanding exactly what this means.

Does that mean: "if the value is True, then the value should be extracted from the regex capture group of the same name" ?  While "if the value is a `str`, then the str value should be used as-is" ?

If this is the meaning, I'm wondering why we need the `true` values.  Is the presence of the named capture groups not enough?  e.g. are there cases where there is a capture group named `type` but you don't want to use its value as `type` later on?  If so, why capture it?

Also, I think there should be at least one layout in the sample data which is using the `str` case.  Right now they are all `true`.

Reply:
> Does that mean: "if the value is True, then the value should be extracted from the regex capture group of the same name" ? While "if the value is a str, then the str value should be used as-is" ?

Yes

> If this is the meaning, I'm wondering why we need the true values. Is the presence of the named capture groups not enough? e.g. are there cases where there is a capture group named type but you don't want to use its value as type later on? If so, why capture it?

Thats a good question. The only cases I see are ChannelDumpsLayout where major is hardcoded ("6") and RhelAlt layouts where platform is hardcoded ("rhel"), and the patterns written for these do capture those segments. I'm not sure why but am trying to keep things as similar as possible to how it was written in cdn-utils.
That aside, I chose this scheme for its flexibility; it allows us to set/unset values manually if needed, despite what might be captured by the regex pattern. Again, not sure of the use case for that but it seemed better.

Reply:
I think it would be cleaner to drop the concept of `True means take the value out of a capture group`.
There is already one special case `...except for rhelalt, where True just means True`.
There could be other boolean attributes to be introduced in the future, leading to more special cases.

I would rather suggest this behavior:

- for each attribute, the default value is `null`.
- a value of `null` means that a capture group of the same name should be used, when present.

Also some of the attributes are probably mandatory, meaning that a value has to be provider either directly (non-null attribute in the layout), or from a capture group, or else a fatal error occurs.  We should somehow document which of them will be mandatory in this way.

Reply:
Adopted this as well. A "true" value is now only necessary when defining rhelalt directory listings (if that value is even valuable, some RhelAlt layouts use it, others don't, and it looks like it's only used once to clear releasevers in ProductContnetSet).

As for mandatory attributes, a scan through cdsutils shows platform and platform_major_version are rather important but I don't think their absence would cause a fatal error. It looks like directory layout attributes just open code paths and aren't strictly required by anything. Either way, I think it would be good to at least log to debug the objects filled when a path is matched.